// css-simple-parser/dist/parse.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
var tokenizer_1 = require("./tokenizer");
var types_1 = require("./types");
var SELECTOR = types_1.TOKEN_TYPE.SELECTOR, BODY_START = types_1.TOKEN_TYPE.BODY_START, BODY_END = types_1.TOKEN_TYPE.BODY_END;
/* GET NODE BODY */
function getNodeBody(node, css) {
    var children = node.children;
    var body = '', start = node.bodyIndex;
    for (var i = 0, l = children.length; i < l; i++) {
        var child = children[i];
        body += css.slice(start, child.index);
        start = child.indexEnd;
    }
    body += css.slice(start, node.bodyIndexEnd);
    return body;
}
/* PARSE */
function parse(css) {
    var tokens = tokenizer_1.default(css), AST = { parent: null, children: [] };
    var parent = AST, index = 0;
    while (true) {
        if (!parent)
            throw new Error('Parent node not found');
        var token = tokens[index];
        if (!token)
            break;
        if (token.type === SELECTOR) {
            var tokenBodyStart = tokens[index + 1];
            if (!tokenBodyStart || tokenBodyStart.type !== BODY_START)
                throw new Error('Found "selector" token without expected subsequent "body_start" token');
            var node = {
                parent: parent,
                index: token.index,
                indexEnd: -1,
                selector: token.selector,
                selectorIndex: token.index,
                selectorIndexEnd: token.indexEnd,
                body: '',
                bodyIndex: tokenBodyStart.index,
                bodyIndexEnd: -1,
                children: []
            };
            parent.children.push(node);
            parent = node;
            index += 2;
        }
        else if (token.type === BODY_END) {
            var node = parent; //TSC
            node.indexEnd = token.index + 1;
            node.bodyIndexEnd = token.index;
            node.body = getNodeBody(node, css);
            parent = node.parent;
            index += 1;
        }
        else {
            throw new Error("Unexpected token of type: \"" + token.type + "\"");
        }
    }
    return AST;
}
/* EXPORT */
exports.default = parse;
